returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,251 @@
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
+
+    def _lazy_init_num_outputs(self):
+        if self.num_outputs is not None:
+            return
+
+        dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self.labels = dataset.labels
+        self._default_input = (
+            "data" if not self._map_outputs or "data" in self._map_outputs else next(self._map_outputs.keys())
+        )
+        self._estimated_num_seqs = dataset.estimated_num_seqs
+
+        if self._map_outputs is not None:
+            tdict = TensorDict(self._map_outputs)
+            self.num_outputs = {
+                k: (t.sparse_dim.size if t.sparse_dim else t.shape[-1] or 1, t.ndim) for k, t in tdict.data.items()
+            }
+            self.num_inputs = self.num_outputs[self._default_input][0]
+        else:
+            self.num_inputs = dataset.num_inputs
+            self.num_outputs = dataset.num_outputs
+
+    def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=None):
+        """
+        :param epoch:
+        :param seq_list:
+        :param seq_order:
+        :return: whether the order changed (True is always safe to return)
+        """
+        super().init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+
+        if epoch is None:
+            self._num_seqs = 0
+            return True
+
+        if self._dataset is None:
+            self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self._dataset.init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+        self._data_iter = enumerate(self._build_dataset_iter(self._dataset))
+        self._data_iter_seq_idx = -1
+        self._lazy_init_num_outputs()
+        return True
+
+    def _collect_single_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        assert seq_idx > self._data_iter_seq_idx, "collecting seqs must be done strictly monotonically"
+        self._data_iter_seq_idx = seq_idx
+
+        while True:
+            try:
+                loaded_seq_idx, tensor_dict = next(self._data_iter)
+            except StopIteration:
+                return None
+            if loaded_seq_idx != seq_idx:
+                continue
+            seq = DatasetSeq(features=tensor_dict.as_raw_tensor_dict(exclude_sequence_lengths=True), seq_idx=seq_idx)
+            return seq
+
+    def _build_dataset_iter(self, dataset: Dataset) -> Iterator[TensorDict]:
+        """
+        :return: an iterator applying both the segment level and across-segment transformations on the given dataset
+        """
+
+        def _wrap_map_seq():
+            for item in self._iterate_dataset(dataset):
+                mapped_item = self._map_seq(item)
+                if isinstance(mapped_item, Iterator):
+                    yield from mapped_item
+                else:
+                    yield mapped_item
```

Comment:
I don't understand how you could get the first case?
```suggestion
                yield mapped_item
```

Reply:
Well, depends on how map_seq is defined we could allow it to return an iterator.

Reply:
We could allow for a million different things. But this seems here very arbitrary and not necessary and not what was discussed before. Let's first only implement the minimal required features that we discussed and make it as simple as possible, definitely not add any complexity like this here. If it turns out that this is a feature some people want, then we can add it later.

Reply:
Ok, that makes sense to me as well!